gem/impero_utils.py

```diff
@@ -316,3 +317,80 @@ def recurse_top(o):
         raise AssertionError("unhandled operation: %s" % type(op))
 
     return counter
+
+
+maxtype = partial(numpy.find_common_type, [])
+
+
+@singledispatch
+def _assign_dtype(expression, self):
+    return maxtype(map(self, expression.children))
+
+
+@_assign_dtype.register(gem.Terminal)
+def _assign_dtype_terminal(expression, self):
+    return self.scalar_type
+
+
+@_assign_dtype.register(gem.Zero)
+@_assign_dtype.register(gem.Identity)
+@_assign_dtype.register(gem.Delta)
+def _assign_dtype_real(expression, self):
+    return self.real_type
+
+
+@_assign_dtype.register(gem.Literal)
+def _assign_dtype_identity(expression, self):
+    return expression.array.dtype
+
+
+@_assign_dtype.register(gem.Power)
+def _assign_dtype_power(expression, self):
+    # Conservative
+    return self.scalar_type
+
+
+@_assign_dtype.register(gem.MathFunction)
+def _assign_dtype_mathfunction(expression, self):
+    if expression.name in {"real", "imag"}:
+        return self.real_type
+    else:
+        return maxtype(map(self, expression.children))
+
+
+@_assign_dtype.register(gem.MinValue)
+@_assign_dtype.register(gem.MaxValue)
+def _assign_dtype_minmax(expression, self):
+    # UFL did correctness checking
+    return self.real_type
+
+
+@_assign_dtype.register(gem.Conditional)
+def _assign_dtype_conditional(expression, self):
+    return maxtype(map(self, expression.children[1:]))
+
+
+@_assign_dtype.register(gem.Comparison)
+@_assign_dtype.register(gem.LogicalNot)
+@_assign_dtype.register(gem.LogicalAnd)
+@_assign_dtype.register(gem.LogicalOr)
+def _assign_dtype_logical(expression, self):
+    return numpy.int8
+
+
+def assign_dtypes(expressions, scalar_type):
+    """Assign numpy data types to expressions.
+
+    Used for declaring temporaries when converting from Impero to lower level code.
+
+    :arg expressions: List of GEM expressions.
+    :arg scalar_type: Default scalar type.
+
+    :returns: list of tuples (expression, dtype)."""
+    mapper = Memoizer(_assign_dtype)
+    mapper.scalar_type = scalar_type
+    if scalar_type.kind == "c":
+        mapper.real_type = scalar_type.type(0).real.dtype
+    else:
+        mapper.real_type = scalar_type
+    return [(e, mapper(e)) for e in expressions]
```

Comment:
This only needs GEM and numpy, but not the Impero AST.  Nevertheless, since it's so closely coupled to code generation, I think it's fine that this visitor stays here.

Reply:
I moved it close to the mathtable in loopy.py now.